packages/astro/src/core/errors/overlay.ts

```diff
@@ -0,0 +1,531 @@
+		this.root.innerHTML = overlayTemplate;
+
+		this.text('#name', err.name);
+		this.text('#title', err.title);
+		this.text('#message-content', err.message, true);
+
+		if (err.hint) {
+			this.text('#hint-content', err.hint, true);
+			this.root.querySelector<HTMLDivElement>('#hint')!.style.display = 'flex';
+		}
+
+		if (err.docslink) {
+			this.root
+				.querySelector('#message')!
+				.appendChild(this.createLink(`See Docs Reference${openNewWindowIcon}`, err.docslink));
+		}
+
+		if (err.loc) {
+			const codeHeader = this.root.querySelector('#code header');
+
+			if (codeHeader && err.loc.file) {
+				const cleanFile = err.loc.file.split('/').slice(-2).join('/');
+				const fileLocation = [cleanFile, err.loc.line, err.loc.column].filter(Boolean).join(':');
+				const absoluteFileLocation = [err.loc.file, err.loc.line, err.loc.column]
+					.filter(Boolean)
+					.join(':');
+
+				const codeFile = codeHeader.getElementsByTagName('h2')[0];
+				codeFile.textContent = fileLocation;
+				codeFile.title = absoluteFileLocation;
+
+				const editorLink = this.createLink(`Open in editor${openNewWindowIcon}`, undefined);
+				editorLink.onclick = () => {
+					fetch('/__open-in-editor?file=' + encodeURIComponent(absoluteFileLocation));
+				};
+
+				codeHeader.appendChild(editorLink);
+			}
+		}
+
+		if (err.highlightedCode) {
+			const codeContent = this.root.querySelector('#code-content');
+
+			if (codeContent) {
+				codeContent.innerHTML = err.highlightedCode;
+
+				window.requestAnimationFrame(() => {
+					const errorLine = this.root.querySelector('.error-line');
+
+					if (errorLine) {
+						errorLine.scrollIntoView(true);
+
+						if (err.loc.column) {
+							errorLine.insertAdjacentHTML(
+								'afterend',
+								`\n<span class="line error-caret"><span style="padding-left:${
+									err.loc.column - 1
+								}ch;">^</span></span>`
+							);
+						}
+					}
+				});
+			}
+		}
+
+		this.text('#stack-content', err.stack);
+	}
+
+	text(selector: string, text: string | undefined, html = false) {
+		if (!text) {
+			return;
+		}
+
+		const el = this.root.querySelector(selector);
+
+		if (el) {
+			if (!html) {
+				el.textContent = text.trim();
+			} else {
+				el.innerHTML = text.trim();
+			}
+		}
+	}
+
+	createLink(text: string, href: string | undefined) {
+		const linkContainer = document.createElement('div');
+		const linkElement = document.createElement('a');
+		linkElement.innerHTML = text;
+
+		if (href) {
+			linkElement.href = href;
+		}
+
+		linkContainer.appendChild(linkElement);
+		linkContainer.className = 'link';
+
+		return linkContainer;
+	}
+}
+
+function getOverlayCode() {
+	return `
+		const overlayTemplate = \`${overlayTemplate}\`;
+		const openNewWindowIcon = \`${openNewWindowIcon}\`;
+
+		const { HTMLElement = class {} } = globalThis;
+		${ErrorOverlay.toString()}
+	`;
+}
+
+export function patchOverlay(code: string) {
+	return (
+		getOverlayCode() +
+		code.replace(
```

Comment:
I can't remember, did we look and see if there was another way to provide a custom overlay? Did that not work for some reason?

Reply:
I investigated taking over the entire client, but since Vite has a lot of code there (mostly related to HMR), we'd need to redo a lot of parts when we really just want to override the overlay. So I figured it wasn't worth it

I changed some code around this part right before sending it in, we could definitely make the replace more reliable

Reply:
Ok, I was under the impression you could configure an alternative custom element constructor and not need to patch the code at all, but looking now I don't see this.

Reply:
I think what this PR does now might be the best way to get complete control of it. Vite could perhaps expose `vite/overlay` in the future to export some utilities, which could be done if more frameworks want to implement their own error overlay.